File: frontend/src/services/adminService.ts
// ============================================================
// services/adminService.ts
// Các API dành riêng cho Role ADMIN.
// Axios tự động đính kèm token.
// ============================================================

import axiosClient from '../api/axiosClient'
import type { ApiResponse, Room } from '../types'

const AdminService = {
  /**
   * GET /api/admin/rooms
   * Lấy tất cả phòng trong hệ thống.
   * Cấp admin có thể xem phòng PENDING, APPROVED, REJECTED kể cả bị ẩn.
   */
  getAllRooms: async (status?: 'PENDING' | 'APPROVED' | 'REJECTED'): Promise<Room[]> => {
    const res = await axiosClient.get<ApiResponse<Room[]>>('/api/admin/rooms', {
      params: { status }
    })
    return res.data.data
  },

  /**
   * PUT /api/admin/rooms/:id/status
   * Duyệt (APPROVED) hoặc từ chối (REJECTED) phòng.
   */
  updateRoomStatus: async (id: number, status: 'APPROVED' | 'REJECTED'): Promise<Room> => {
    const res = await axiosClient.put<ApiResponse<Room>>(`/api/admin/rooms/${id}/status`, { status })
    return res.data.data
  },

  /**
   * DELETE /api/admin/rooms/:id
   * Xóa vĩnh viễn phòng (kể cả của landlord khác).
   */
  deleteRoom: async (id: number): Promise<void> => {
    await axiosClient.delete(`/api/admin/rooms/${id}`)
  },
  
  /**
   * PATCH /api/admin/rooms/:id/availability
   * Toggle Còn phòng / Hết phòng.
   */
  toggleAvailability: async (id: number): Promise<Room> => {
    const res = await axiosClient.patch<ApiResponse<Room>>(`/api/admin/rooms/${id}/availability`)
    return res.data.data
  },
}

export default AdminService

File: src/main/java/com/tronhanh/service/AdminService.java
package com.tronhanh.service;

import com.tronhanh.dto.response.RoomResponse;
import com.tronhanh.dto.response.UserResponse;
import com.tronhanh.entity.RoomStatus;

import java.util.List;

/**
 * Interface nghiệp vụ quản trị hệ thống dành cho Admin.
 */
public interface AdminService {

    /**
     * Lấy tất cả phòng trọ trong hệ thống (Admin xem toàn bộ).
     *
     * @param status null = lấy tất cả; PENDING/APPROVED/REJECTED = lọc theo trạng thái
     */
    List<RoomResponse> getAllRooms(RoomStatus status);

    /**
     * Cập nhật trạng thái phê duyệt phòng.
     * Chỉ nhận APPROVED hoặc REJECTED — không được set PENDING.
     */
    RoomResponse updateRoomStatus(Long roomId, RoomStatus status);

    /**
     * Xóa phòng vi phạm (Admin có thể xóa bất kỳ phòng nào).
     * Cascade xóa saved_rooms và room_images.
     */
    void deleteRoom(Long roomId);

    /**
     * Toggle khóa/mở khóa tài khoản người dùng.
     * Tài khoản bị khóa không thể đăng nhập.
     */
    UserResponse toggleUserLock(Long userId);

    /**
     * Lấy tất cả người dùng trong hệ thống.
     * Optional filter: ?role=RENTER | LANDLORD | ADMIN
     */
    List<UserResponse> getAllUsers(String role);

    /**
     * Bật/tắt trạng thái còn phòng/hết phòng.
     */
    RoomResponse toggleAvailability(Long roomId);
}
